enhancements/cluster-logging/elasticsearch-hot-warm-architecture.md

```diff
@@ -0,0 +1,156 @@
+
+### Non-Goals
+
+* We will not be marking different nodes for the different policies and reallocating shards based on that
+
+* We will not be implementing this using the Opendistro plugin
+
+## Proposal
+
+Currently we offer a mechanism for users to configure their index lifecycles based on two states: hot and delete. By default indices are created and stay in the hot state, where they can be rolled over based on a configuration, however they stay in a "hot" state until they are eventually deleted based on the delete policy.
+
+Elasticsearch offers mechanisms to move indices between hot, warm, cold, and delete and discusses approaches for this [in a blog article](https://www.elastic.co/blog/implementing-hot-warm-cold-in-elasticsearch-with-index-lifecycle-management). However, this functionality is a x-pack feature and we do not support these. [Opendistro](https://opendistro.github.io/for-elasticsearch-docs/docs/ism/) offers a way to do this as well, however their plugins require ES7. We achieve our current functionality using cronjobs -- one for delete and one for rollover.
+
+Within the context of Elasticsearch ILM:
+1) Moving the index into a Warm state means the index is force merged to use less primary shards, its replica count is reduced, and the index priority is also reduced. The idea of this being that these indices are not being written to, may not be queried against as much so they do not need to be as distributed, and can be a lower priority when it comes to index recovery (meaning they would be recovered after high priority indices are).
+
+1) Moving into a Cold state further reduces its index priority to be lower than Warm (but still higher than delete) and, if one was using specific nodes to house indices of certain state it would move from Warm to Cold nodes. This is different than Closing an index because the index can still be queried (if one were to Close an index it would need to be Opened first before it can be searched).
+
+### User Stories
+
+#### As an EO admin, I want to be able to configure a warm policy for my indices so that I can keep searchable data around longer and take up less resources
+
+Currently we keep indices at their original primary and replica shard count which comes at a cost in ES. Reducing these can help us increase our retention policy.
+
+### Implementation Details
+
+#### Assumptions
+
+* This will be going out with our ES6 operator which means it will be a part of our ES 6.x cluster
+
+* If not specified to be separate Warm and Rollover will be synonomous
+
+#### Security
+
+No additional security concerns need to be addressed by this.
+
+### Risks and Mitigations
+
+## Design Details
+
+Currently the API reads differently for Hot and Delete. Hot uses the MaxAge of an index and Delete the MinAge. For the sake of this proposal Warm will follow the same convention as Delete, a later iteration can seek to align all phases to use the same terminology (non-goal).
```

Comment:
Is it useful to have some validation here? e.g. `warm.minAge > cold.minAge > delete.minAge`? I assume users want to configure something like: `warm = 30d, cold = 60 days, delete = 90d`

Reply:
We would only have `hot`, `warm`, and `delete` phases using ISM (I do not feel `cold` buys us much of anything vs `warm` given our architectural offering).

I'm not sure its worth us doing validation, if a customer wants to configure it such that warm would not do anything they should be allowed to do so.

Reply:
What is the load impact on Elasticsearch, if the three states are configured too eager, e.g. the different by one day? Does eager reduction of replicas and shards require a lot of resources from ES?

Reply:
as far as I understand, reducing replicas is a pretty low effort action since there is no need to sync data we're simply removing things. I'm not sure how the force merge effects the cluster, but this is something ES recommends to do so I would imagine they have made it performant enough cc @lukas-vlcek